File: 3393-count-paths-with-the-given-xor-value/3393-count-paths-with-the-given-xor-value.java
class Solution {
    long mod=1000000007;
    public int countPathsWithXorValue(int[][] grid, int k) {
        int n=grid.length;
        int m=grid[0].length;
        long[][][] ways=new long[n][m][16];
        ways[0][0][grid[0][0]]=1;
        int t=grid[0][0];
        for (int i=1;i<n;i++){
            t=t^grid[i][0];
            ways[i][0][t]=1;
        }
        t=grid[0][0];
        for (int i=1;i<m;i++){
            t=t^grid[0][i];ways[0][i][t]=1;
        }
        for (int i=1;i<n;i++){
            for (int j=1;j<m;j++){
                for ( t=0;t<16;t++){
                    int t1=t^grid[i][j];
                    ways[i][j][t]=(ways[i-1][j][t1]+ways[i][j-1][t1])%mod;
                }
            }
        }
        return (int)ways[n-1][m-1][k];
    }
}

File: 3393-count-paths-with-the-given-xor-value/3393-count-paths-with-the-given-xor-value.py
from typing import List
class Solution:
    def countPathsWithXorValue(self, grid: List[List[int]], k: int) -> int:
        n = len(grid)
        m = len(grid[0])
        ans = [[0] * 16]
        ans[0][grid[0][0]] = 1
        divisor = 1000000007
        for i in range(1, m):
            tmp = []
            for j in range(16):
                tmp.append(ans[i - 1][grid[0][i] ^ j])
            ans.append(tmp)
        for i in range(1, n):
            tmp = [[]]
            for j in range(16):
                tmp[0].append(ans[0][grid[i][0] ^ j])
            for j in range(1, m):
                now = []
                for kk in range(16):
                    target = kk ^ grid[i][j]
                    now.append((tmp[-1][target] + ans[j][target]) % divisor)
                tmp.append(now)
            ans = tmp
        return ans[-1][k]